libnd4j/include/ops/declarable/generic/helpers/impl/convolutions.cpp

```diff
@@ -415,9 +415,161 @@ else
 }
 
 
+#ifdef HAVE_MKLDNN
+using namespace mkldnn;
+
+void ConvolutionUtils::getMKLDNNMemoryDescConv2d(
+        int kH, int kW, int sH, int sW, int pH, int pW, int dH, int dW, bool isSameMode, bool isNCHW,
+        int bS, int iC, int iH, int iW, int oC, int oH, int oW, const NDArray* src, const NDArray* diff_src,
+        const NDArray* weights, const NDArray* diff_weights, const NDArray* bias, const NDArray* dst,
+        mkldnn::memory::desc* conv_src_md, mkldnn::memory::desc* conv_diff_src_md, mkldnn::memory::desc* conv_weights_md,
+        mkldnn::memory::desc* conv_diff_weights_md, mkldnn::memory::desc* conv_bias_md, mkldnn::memory::desc* conv_dst_md,
+        mkldnn::memory::dims& conv_strides, mkldnn::memory::dims& conv_padding, mkldnn::memory::dims& conv_padding_r) {
+    mkldnn::memory::dims conv_src_tz = { bS, iC, iH, iW };
+    mkldnn::memory::dims conv_weights_tz = { oC, iC, kH, kW };
+    mkldnn::memory::dims conv_bias_tz = { oC };
+    mkldnn::memory::dims conv_dst_tz = { bS, oC, oH, oW };
+
+    conv_strides = { sH, sW };
+    conv_padding = { pH, pW };
+    conv_padding_r = { (oH - 1) * sH - iH + kH - pH,
+                       (oW - 1) * sW - iW + kW - pW };
```

Comment:
conv_padding is 'truncate' mode padding values, and conv_padding_r is same mode padding, right?
If so, not sure on the -pH and -pW here... Usually with same mode, we ignore the ph and pW args and just calculate what we should actually use.
https://github.com/deeplearning4j/deeplearning4j/blob/master/deeplearning4j/deeplearning4j-nn/src/main/java/org/deeplearning4j/nn/conf/ConvolutionMode.java#L62-L63

Also present in a bunch of other places.

Reply:
No, conv_padding_r is the padding on the "right", but it needs to match that equation exactly or MKL-DNN will complain: https://github.com/intel/mkl-dnn/blob/master/src/common/convolution.cpp#L116